### ISSUE TITLE
Help with Better player overlay 

### ISSUE THREAD

**Issue by febg11** [jhomlala/betterplayer#208]
This isnt really a bug I am just struggling with 2 problems. The first is  setting up an overlay that has values based on the _betterPlayerController.

For example I would like to have a custom play/pause button that switches when the user presses one. I thought the overlay would be best for this as it will also work in the expanded mode.

However when i do this 

```
    _betterPlayerController = new BetterPlayerController(
      BetterPlayerConfiguration(
        fullScreenAspectRatio: aspectRatio,
        aspectRatio: aspectRatio,
        overlay: buildOverlay(),
        controlsConfiguration: BetterPlayerControlsConfiguration(
          showControls: false
        ),
        eventListener: (BetterPlayerEvent event) {
          if(event.betterPlayerEventType == BetterPlayerEventType.initialized) {
            setState(() {});
          }
        },
      ),


  Widget buildOverlay() {
        
    if(_betterPlayerController == null ){
      return Container(color: Colors.yellow,);
    }
  .....
}
```
It will always return the yellow background, I assume its because the method is called only once, when we construct the BetterPlayerConfiguration().

My second issue relates to how my widget works. I pass in either a string (webURL) or local file. If a webURL is passed the file will be downloaded before being placed in the video. This has led to a strange FutureBuilder and call in the build to initialise the file, alongside a didUpdateWidget to reinitialise if either of the inputs change. 

```

  @override
  void initState() {
    super.initState();
    initialiseFuture(url: widget.url false);
  }

  @override
  didUpdateWidget(BetterPlayerVideo old) {
    super.didUpdateWidget(old);

    // initialiseFuture will initialise downloadAndCacheFile
    if (widget.file != null && widget.file != old.file) {
      initialiseFuture(url: null, false);
    } else if (widget.url != null && widget.url != old.url) {
      initialiseFuture(url: widget.url, false);
    }
  }


  @override
  Widget build(BuildContext context) {
    return Builder(builder: (BuildContext context) {
      ctx= context;

      if (widget.file != null) {
        tryUpdateDataSource(widget.file);
        return getVideoPlayer();
      } else {
        return FutureBuilder(
            future: downloadAndCacheFile,
            builder: (BuildContext context, AsyncSnapshot<File> cachedFile) {
              if (cachedFile.hasData) {
                tryUpdateDataSource(cachedFile.data);
                return getVideoPlayer();
              } else {
                return Text('Loading');
              }
            });
      }
    });
  }

  tryUpdateDataSource(File file) {
    if (_betterPlayerController == null || _betterPlayerController.betterPlayerDataSource == null || _betterPlayerController.betterPlayerDataSource.url != file.path) {
      BetterPlayerDataSource dataSource = BetterPlayerDataSource.file(file.path);
      if (_betterPlayerController == null) {
        setUpController(dataSource);
      }else{
        _betterPlayerController.setupDataSource(dataSource);

      }
      buildLocalControls();
    } else {
      print('Skipping as file is the same');
    }
  }


  Widget getVideoPlayer() {
    return Container(
      child: BetterPlayer(
        controller: _betterPlayerController,
      ),
    );
  }
```

I really feel I am approaching this completely wrong (and using the library incorrectly), as it checks whether we need to initialise every build. I would appreciate any help anyone can provide.

**Comment by jhomlala** [comment]
@febg11 
1st issue:
```dart
import 'dart:async';

import 'package:better_player/better_player.dart';
import 'package:better_player_example/constants.dart';
import 'package:flutter/material.dart';

class ExamplePage extends StatefulWidget {
  @override
  _ExamplePageState createState() => _ExamplePageState();
}

class _ExamplePageState extends State<ExamplePage> {
  BetterPlayerController _betterPlayerController;

  ///Stream controller which updates on play/pause changes
  StreamController<bool> _isPlayingStreamController =
      StreamController.broadcast();

  @override
  void initState() {
    BetterPlayerConfiguration betterPlayerConfiguration =
        BetterPlayerConfiguration(
      aspectRatio: 16 / 9,
      fit: BoxFit.contain,
          overlay: _buildOverlay(),
    );
    _betterPlayerController = BetterPlayerController(betterPlayerConfiguration);
    _betterPlayerController.addEventsListener((event) {
      if (event.betterPlayerEventType == BetterPlayerEventType.play) {
        _isPlayingStreamController.add(true);
      }
      if (event.betterPlayerEventType == BetterPlayerEventType.pause) {
        _isPlayingStreamController.add(false);
      }
    });
    _setupDataSource();
    super.initState();
  }

  void _setupDataSource() async {
    BetterPlayerDataSource dataSource = BetterPlayerDataSource(
      BetterPlayerDataSourceType.network,
      Constants.elephantDreamVideoUrl,
    );
    _betterPlayerController.setupDataSource(dataSource);
  }

  @override
  void dispose() {
    _isPlayingStreamController.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Notification player"),
      ),
      body: Column(
        children: [
          AspectRatio(
            aspectRatio: 16 / 9,
            child: BetterPlayer(controller: _betterPlayerController),
          ),
        ],
      ),
    );
  }

  Widget _buildOverlay() {
    return Center(
      child: StreamBuilder<bool>(
        stream: _isPlayingStreamController.stream,
        builder: (context, snapshot) {
          bool isPlaying = snapshot.data ?? false;
          if (isPlaying) {
            return Text(
              "Playing",
              style: TextStyle(color: Colors.yellow),
            );
          } else {
            return Text(
              "Paused",
              style: TextStyle(color: Colors.red),
            );
          }
        },
      ),
    );
  }
}

```

Use event listener to catch BetterPlayer events. Once proper event has been catched, you can use StreamController and StreamBuilder to update UI without re-rendering whole widget tree.

2nd problem:
What you need to do is create StreamBuilder and StreamController. StreamController holds url of local file. Once file has been downloaded (and StreamController has been updated), you can show BetterPlayer in StreamBuilder. This is a pseudo code:

```///Url of the local file (which is on device)
StreamController<String> fileUrlStreamController
BetterPlayerController betterPlayerController;

void initState(){
	processFile();
	betterPlayerController = ///Init betterPlayerController
}

void processFile(String url){
if (url is network url){
		String localUrl = downloadFile(url)
		fileDownloaded.add(url)
	} else {
		fileDownloaded.add(url);
	}
}


String downloadFile(String url){
	///Download file sand save it to the device drive
	///Return url of downloaded file
}

Widget buildPlayer(){
	return StreamBuilder(
		stream: fileUrlStreamController
		builder:(snapshot,context){
			var url = snapshot.data;
			if (url != null){
				betterPlayerController.setupDataSource(url);
				return BetterPlayer(betterPlayerController : betterPlayerController)
			} else {
				///Show placeholder widget
				return Text("Loading");
			}
			
		
		})
}

Widget build(){
	return Scaffold(body: AspectRatio(buildPlayer()));
}```

**Comment by febg11** [comment]
Thanks so much, your approach made it a lot simpler, i have just implemented. I was originally using the Video Player library and the refactor to yours has made it neater but has led to a few things that which may not be possible using the library, I cant immediately see a way to do them.

Could you clarify whether these things are possible

1) When the video is fullscreen, allowing the overlay to fill the entire screen (e.g. expanding past the black bars) 
2) Having the ability to Make the full screen mode fill the entier screen (the flutter debug label seems to stop below the top of by screen)
3) Ability to change the background behind the expanded video... e.g. put a Widget that will appear above and below the video instead of the black bars. 
4) Is it possible for the full screen mode to still show the notification bar rather than hiding it?
5) Change the transition when the video is expanded... e.g. to the navigator.push()

I hope its possible to do these things :)

**Comment by jhomlala** [comment]
Answer for your question is yes - it's should be doable with custom route. 
Bug reported here: can you create new issue with reproducible code which can be run? Thanks!
Anyway I'll close this task since original issue has been fixed. 
Feel free to ask questions in discussions page here: https://github.com/jhomlala/betterplayer/discussions.
